### PR TITLE
Improvement/convert utils and tracker to ts

### PR DESCRIPTION
**Pull request by ethan605**
# Problem

This is a prerequisite for #1315

## Checklist

_put `n/a` if item is not relevant to PR changes_

- [x] Has the CHANGELOG been updated?
- [n/a] Has the README been updated?
- [n/a] Has the CONTRIBUTING doc been updated?
- [n/a] Has the RELEASE_GUIDELINES been updated?
- [n/a] Has the MIGRATION doc been updated for any MAJOR breaking changes?
- [n/a] Has the MIGRATION doc been updated for any MINOR breaking changes, including any translation strings or keys changes?
- [n/a] Have any new automated tests been implemented or the existing ones changed?
- [n/a] Have any new manual tests been written down or the existing ones changed?
- [n/a] Have any new strings been translated or the existing ones changed?

### CODE REVIEW

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/utils/camera.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+export const getRecordedVideo = (
+  webcam: Webcam,
+  callback: HandleCaptureProp
+): void => {
+  const blob = webcam.getVideoBlob()
+  const sdkMetadata = getDeviceInfo(webcam.stream)
+  callback({ blob, sdkMetadata })
+}
+
+const getDeviceInfo = (stream: MediaStream): SdkMetadata => {
```
Comment (stefaniacardenas) [inline src/components/utils/camera.ts:37]: `SdkMetadata` could be an empty object, does that need to be specified here or is that already expected in the `SdkMetadata` declaration?

File: src/components/utils/index.tsx [hunk 2]
```diff
@@ -1,27 +1,49 @@
...
     console.warn(
       `The css @value: ${cssUnit} unit is ${resUnit} but it should be ${expectedUnit}`
     )
   }
+
   return value
 }
 
-export const getCSSMilisecsValue = (cssUnit) => getCSSValue('ms', cssUnit)
+export const getCSSMilisecsValue = (cssUnit: string | number): number =>
```
Comment (stefaniacardenas) [inline src/components/utils/index.tsx:36]:
```suggestion
export const getCSSMillisecsValue = (cssUnit: string | number): number =>
```

File: src/components/utils/string.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+https://api.onfido.com/v2/documents
+https://onfido.com/v2/documents
+https://cross.onfido.com/v2/documents
+https://cross.lol.onfido.com/v2/documents
+
+Tested fail against:
+https://revolut.com/v2/documents/?url=https://onfido.com", /https:\/\/[A-Za-z0-9\.]*\.?onfido\.com/g
+https://onfido.revolut.com/v2/documents
+https://onfido.revolut.com/v2/documents/onfido.com
+ */
```
Comment (stefaniacardenas) [inline src/components/utils/string.ts:58]: These comments are out of date I believe because we use v3 endpoints

**Review by stefaniacardenas — APPROVED** [review]

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

File: src/components/utils/camera.ts (hunk 1 above)
Comment (ethan605) [reply src/components/utils/camera.ts:37]: Yes, the declaration of [`SdkMetadata` type](https://github.com/onfido/onfido-sdk-ui/blob/improvement/convert-utils-and-tracker-to-ts/src/types/commons.ts#L27-L40) includes optional fields, so an empty object is a legit `SdkMetadata`

File: src/components/utils/string.ts (hunk 3 above)
Comment (ethan605) [reply src/components/utils/string.ts:58]: Great, thanks for the confirmation!

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

File: src/supported-documents/index.ts [hunk 4]
```diff
@@ -1,14 +1,21 @@
-import supportedDrivingLicences from './supported-docs-driving_licence.json'
-import supportedNationalIDCards from './supported-docs-national_identity_card.json'
-import supportedResidencePermit from './supported-docs-residence_permit.json'
+/* eslint-disable @typescript-eslint/no-var-requires */
+const supportedDrivingLicences = require('./supported-docs-driving_licence.json')
+const supportedNationalIDCards = require('./supported-docs-national_identity_card.json')
+const supportedResidencePermit = require('./supported-docs-residence_permit.json')
```
Comment (kopijunkie) [inline src/supported-documents/index.ts:4]: No strong feelings about this, but just curious why this change was needed since an eslint rule also needs to be disabled because of it?

**Review by kopijunkie — APPROVED** [review]

File: src/Tracker/index.tsx [hunk 5]
```diff
@@ -92,32 +98,45 @@ const install = () => {
 
       const shouldReturnCrumb = isOnfidoXhr || isOnfidoClick
 
-      return shouldReturnCrumb ? crumb : false
+      return shouldReturnCrumb ? crumb : null
     },
-    shouldSendCallback: () => process.env.PRODUCTION_BUILD,
+    // @TODO: verify these mismatched options
+    // autoBreadcrumbs: { console: false },
+    // shouldSendCallback: () => process.env.PRODUCTION_BUILD,
```
[the remark this answers is not in the document]
Comment (stefaniacardenas) [reply src/Tracker/index.tsx:105]: This is legacy from when we were using `raven`, an older sentry package. They should be removed. I created a ticket a while ago to clean this up

File: src/Tracker/index.tsx [hunk 6]
```diff
@@ -144,48 +173,60 @@ const appendToTracking = (Acomponent, ancestorScreeNameHierarchy) =>
...
       this.props.trackScreen(screenName)
     }
-    render = () => <Acomponent {...this.props} />
+
+    render = () => <WrappedComponent {...this.props} />
   }
 
-const trackComponentMode = (Acomponent, propKey) =>
-  class TrackedComponentWithMode extends Component {
+const trackComponentMode = <P extends WithTrackingProps>(
```
[the remark this answers is not in the document]
Comment (ethan605) [reply src/Tracker/index.tsx:193]: Oh that's TypeScript's [Generics syntax](https://www.typescriptlang.org/docs/handbook/generics.html)

File: src/supported-documents/index.ts (hunk 4 above)
Comment (ethan605) [reply src/supported-documents/index.ts:4]: By default, TypeScript should properly recognise JSON modules imported by ES6 imports (`import * from 'something.json'`). But I got errors when using it within our project, so just using `require` imports here as a workaround and try to sort it out later

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://staging-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://release-1312-pr-onfido-sdk-ui-onfido.surge.sh

**Comment by rfreitas** [comment]
Travis automatic deployment: https://1312-pr-onfido-sdk-ui-onfido.surge.sh
